Fix raster legend assertion bug

nkorinek opened the pull request:
@lwasser Here's a PR fixing the issue brought up in https://github.com/earthlab/matplotcheck/issues/163 . While my last attempt lead to me thinking this issue was unsolvable, I reviewed my old code and realized I was going about it in the entirely wrong way! This solution is a lot more elegant and efficient, and it seems to work! If I need to add in any tests to go alongside this please let me know. Thanks!

nkorinek committed:
Merge pull request #1 from earthlab/master

Updating files

pyup-bot committed:
# This is a combination of 2 commits.
# This is the 1st commit message:

Update setuptools from 42.0.2 to 45.2.0 (#190)

raster inherits from vector rather than base (#76)

Allow geodataframes in assert_polygons (#188)

fixing a few syntax errors (#194)

* First draft of get_images function

* Added a test for get_image

* Changelog update

* minor formatting change.

* black changes

* fixing a few syntax errors

* Update matplotcheck/raster.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* remove tabls

Co-authored-by: Nathan Korinek <nako1890@colorado.edu>

update changelog for release

Bump version: 0.1.1 → 0.1.2

Add flake8 fix [WIP] (#199)

* add flake 8

* yay cleanup

* update change log

* add flake 8 dep

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* more line length issues

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* fixed test_base_axis.py

* fixed text_base_data.py

* fixed test_base_legends.py

* fixed test_base_titles_captions.py

* Format files to be flake8 acceptable (#197)

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* Fixed test_raster.py and issue with legends tests

* black

* more black changes

* Fixing merge conflicts that were found

* Changing accepted changes

* minor changes

* Changing to make CI happy, modified to pass black and make -B docs

* Reformatting to make black and flake8 happy. Also updated crs assignment

* Fixed formatting issue with timeseries

* Better fix for timeseries formatting

* Black

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

Assert string accept spaces (#205)

* Made it so that assert_string_contains accepts key words with spaces

* black

* renaming function

* Showed assert_title_contains working with a space in the keyword in vignette

Update codecov to 2.0.16 (#202)

* Update codecov from 2.0.15 to 2.0.16

* update from master (#211)

* Add flake8 fix [WIP] (#199)

* add flake 8

* yay cleanup

* update change log

* add flake 8 dep

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* more line length issues

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* fixed test_base_axis.py

* fixed text_base_data.py

* fixed test_base_legends.py

* fixed test_base_titles_captions.py

* Format files to be flake8 acceptable (#197)

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* Fixed test_raster.py and issue with legends tests

* black

* more black changes

* Fixing merge conflicts that were found

* Changing accepted changes

* minor changes

* Changing to make CI happy, modified to pass black and make -B docs

* Reformatting to make black and flake8 happy. Also updated crs assignment

* Fixed formatting issue with timeseries

* Better fix for timeseries formatting

* Black

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

* Assert string accept spaces (#205)

* Made it so that assert_string_contains accepts key words with spaces

* black

* renaming function

* Showed assert_title_contains working with a space in the keyword in vignette

Co-authored-by: Nathan Korinek <nako1890@colorado.edu>

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>
Co-authored-by: Nathan Korinek <nako1890@colorado.edu>

Contributors update (#213)

* Added contributor rst file and moved that information out of the README file

* Implemented requested changes to README

Update setuptools from 45.2.0 to 46.0.0 (#215)

Add Assert points function & cleanup duplicate methods (#203)

* Added a get_points() and assert_points() function to the vector tester.

* Added in get_points() and assert_points() functions with tests

* Added proper documentation

* Small fix to please codacy

* black

* Updated changelog

* Rough draft for bug fix

* Fixed bug with multiple geometries plotted alongside bug with identical x values causing failure!

* typo

* Fixed small bug with markersize

* Added comments explaining code

* Taking out redundant tests

* Typo

* Fixing how vector checks for truth value of a dataframe

Update pytest from 5.3.5 to 5.4.1

merge

Additional vector tests (#212)

* Added a get_points() and assert_points() function to the vector tester.

* Added in get_points() and assert_points() functions with tests

* Added proper documentation

* Small fix to please codacy

* black

* Updated changelog

* First round of tests for vector

* black

* Rough draft for bug fix

* Fixed bug with multiple geometries plotted alongside bug with identical x values causing failure!

* typo

* Fixed small bug with markersize

* Added comments explaining code

* rough drafts of more tests

* Added more tests for legends

* Added more tests, and took out broken tests

* small codacy fix

* Fixed test!

* Taking out redundant tests

* Took out unneccesary tests

* Added plt.close

* Added more tests

* Typo

* more tests for uncovered parts of the vector file

* black

* Fixed issues with vector checking truth value of dataframe, and added tests

* Fixing how vector checks for truth value of a dataframe

* Added more coverage!

* Broke tests up into individual files

* black

* Added tests for outlier cases

* Update matplotcheck/tests/test_points.py

typo

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* took plt.gca() out of tests

* Added changes suggested on GitHub

* Update CHANGELOG.md

* fix import order

* import order fix

* import order

* import order

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

Add tests to the Autograde module (#224)

* Added tests for the autograde module

* removed unneccesary import

* Fixing issues that were failing in base_data

* codacy

* flake8

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Updated changelog

* flake8 fix

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

Hist bin midpoints (#204)

* Add a get and assert function for bin midpoints, as well as tests for those functions

* Updated docstrings

* Added changelog changes and fixed minor formatting issue

* small flake8 fix

* Added midpoints functionatlity to the vignette for histogram testing

* Adding in changes suggested on GitHub1

* Fixed function description

* Took out all instances of  in example

* update docs

* Fixed small bug with section titles

* Update matplotcheck/tests/test_base_data.py

* reworded changelog

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

Update setuptools from 46.1.1 to 46.1.3 (#231)

Title assert accepts strings (#229)

* Fixed assert_title_contains to take strings as well as lists

* Better implementation

* Changelog update

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

M2r remove (#247)

* Removed m2r from matplotcheck, and reformated code of conduct

* Seeing if rebase works

* removed an m2r import

* Fixing up the docs a bit

* changelog updates

Allow geodataframes in assert_polygons (#188)

fixing a few syntax errors (#194)

* First draft of get_images function

* Added a test for get_image

* Changelog update

* minor formatting change.

* black changes

* fixing a few syntax errors

* Update matplotcheck/raster.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* remove tabls

Co-authored-by: Nathan Korinek <nako1890@colorado.edu>

Get images function (#193)

* First draft of get_images function

* Added a test for get_image

* Changelog update

* minor formatting change.

* black changes

* Update matplotcheck/raster.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

Bump version: 0.1.1 → 0.1.2

Add flake8 fix [WIP] (#199)

* add flake 8

* yay cleanup

* update change log

* add flake 8 dep

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* more line length issues

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* fixed test_base_axis.py

* fixed text_base_data.py

* fixed test_base_legends.py

* fixed test_base_titles_captions.py

* Format files to be flake8 acceptable (#197)

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* Fixed test_raster.py and issue with legends tests

* black

* more black changes

* Fixing merge conflicts that were found

* Changing accepted changes

* minor changes

* Changing to make CI happy, modified to pass black and make -B docs

* Reformatting to make black and flake8 happy. Also updated crs assignment

* Fixed formatting issue with timeseries

* Better fix for timeseries formatting

* Black

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

Contributors update (#213)

* Added contributor rst file and moved that information out of the README file

* Implemented requested changes to README

Add Assert points function & cleanup duplicate methods (#203)

* Added a get_points() and assert_points() function to the vector tester.

* Added in get_points() and assert_points() functions with tests

* Added proper documentation

* Small fix to please codacy

* black

* Updated changelog

* Rough draft for bug fix

* Fixed bug with multiple geometries plotted alongside bug with identical x values causing failure!

* typo

* Fixed small bug with markersize

* Added comments explaining code

* Taking out redundant tests

* Typo

* Fixing how vector checks for truth value of a dataframe

Additional vector tests (#212)

* Added a get_points() and assert_points() function to the vector tester.

* Added in get_points() and assert_points() functions with tests

* Added proper documentation

* Small fix to please codacy

* black

* Updated changelog

* First round of tests for vector

* black

* Rough draft for bug fix

* Fixed bug with multiple geometries plotted alongside bug with identical x values causing failure!

* typo

* Fixed small bug with markersize

* Added comments explaining code

* rough drafts of more tests

* Added more tests for legends

* Added more tests, and took out broken tests

* small codacy fix

* Fixed test!

* Taking out redundant tests

* Took out unneccesary tests

* Added plt.close

* Added more tests

* Typo

* more tests for uncovered parts of the vector file

* black

* Fixed issues with vector checking truth value of dataframe, and added tests

* Fixing how vector checks for truth value of a dataframe

* Added more coverage!

* Broke tests up into individual files

* black

* Added tests for outlier cases

* Update matplotcheck/tests/test_points.py

typo

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* took plt.gca() out of tests

* Added changes suggested on GitHub

* Update CHANGELOG.md

* fix import order

* import order fix

* import order

* import order

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

# This is the commit message #2:

Update setuptools from 46.1.1 to 46.1.3 (#231)

pyup-bot committed:
Update setuptools from 42.0.2 to 45.2.0 (#190)

nkorinek committed:
Merge pull request #2 from earthlab/master

Update setuptools from 42.0.2 to 45.2.0 (#190)

Commit:
raster inherits from vector rather than base (#76)

nkorinek committed:
Merge pull request #3 from earthlab/master

raster inherits from vector rather than base (#76)

nkorinek committed:
Allow geodataframes in assert_polygons (#188)

nkorinek committed:
Merge pull request #4 from earthlab/master

Allow geodataframes in assert_polygons (#188)

Commit:
fixing a few syntax errors (#194)

* First draft of get_images function

* Added a test for get_image

* Changelog update

* minor formatting change.

* black changes

* fixing a few syntax errors

* Update matplotcheck/raster.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* remove tabls

Co-authored-by: Nathan Korinek <nako1890@colorado.edu>

nkorinek committed:
Merge pull request #5 from earthlab/master

Updating from master

Commit:
update changelog for release

Commit:
Bump version: 0.1.1 → 0.1.2

nkorinek committed:
Merge pull request #7 from earthlab/master

updating

nkorinek committed:
Add flake8 fix [WIP] (#199)

* add flake 8

* yay cleanup

* update change log

* add flake 8 dep

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* more line length issues

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* fixed test_base_axis.py

* fixed text_base_data.py

* fixed test_base_legends.py

* fixed test_base_titles_captions.py

* Format files to be flake8 acceptable (#197)

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* Fixed test_raster.py and issue with legends tests

* black

* more black changes

* Fixing merge conflicts that were found

* Changing accepted changes

* minor changes

* Changing to make CI happy, modified to pass black and make -B docs

* Reformatting to make black and flake8 happy. Also updated crs assignment

* Fixed formatting issue with timeseries

* Better fix for timeseries formatting

* Black

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

nkorinek committed:
Merge pull request #8 from earthlab/master

Add flake8 fix [WIP] (#199)

nkorinek committed:
Assert string accept spaces (#205)

* Made it so that assert_string_contains accepts key words with spaces

* black

* renaming function

* Showed assert_title_contains working with a space in the keyword in vignette

nkorinek committed:
Merge pull request #9 from earthlab/master

Assert string accept spaces (#205)

pyup-bot committed:
Update codecov to 2.0.16 (#202)

* Update codecov from 2.0.15 to 2.0.16

* update from master (#211)

* Add flake8 fix [WIP] (#199)

* add flake 8

* yay cleanup

* update change log

* add flake 8 dep

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* more line length issues

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* fixed test_base_axis.py

* fixed text_base_data.py

* fixed test_base_legends.py

* fixed test_base_titles_captions.py

* Format files to be flake8 acceptable (#197)

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* Fixed test_raster.py and issue with legends tests

* black

* more black changes

* Fixing merge conflicts that were found

* Changing accepted changes

* minor changes

* Changing to make CI happy, modified to pass black and make -B docs

* Reformatting to make black and flake8 happy. Also updated crs assignment

* Fixed formatting issue with timeseries

* Better fix for timeseries formatting

* Black

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

* Assert string accept spaces (#205)

* Made it so that assert_string_contains accepts key words with spaces

* black

* renaming function

* Showed assert_title_contains working with a space in the keyword in vignette

Co-authored-by: Nathan Korinek <nako1890@colorado.edu>

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>
Co-authored-by: Nathan Korinek <nako1890@colorado.edu>

nkorinek committed:
Contributors update (#213)

* Added contributor rst file and moved that information out of the README file

* Implemented requested changes to README

pyup-bot committed:
Update setuptools from 45.2.0 to 46.0.0 (#215)

nkorinek committed:
Merge pull request #10 from earthlab/master

update

pyup-bot committed:
Update pytest from 5.3.5 to 5.4.1

nkorinek committed:
Add Assert points function & cleanup duplicate methods (#203)

* Added a get_points() and assert_points() function to the vector tester.

* Added in get_points() and assert_points() functions with tests

* Added proper documentation

* Small fix to please codacy

* black

* Updated changelog

* Rough draft for bug fix

* Fixed bug with multiple geometries plotted alongside bug with identical x values causing failure!

* typo

* Fixed small bug with markersize

* Added comments explaining code

* Taking out redundant tests

* Typo

* Fixing how vector checks for truth value of a dataframe

nkorinek committed:
Merge pull request #11 from earthlab/master

Add Assert points function & cleanup duplicate methods (#203)

nkorinek committed:
Merge pull request #12 from earthlab/master

Update pytest from 5.3.5 to 5.4.1

nkorinek committed:
Merge pull request #13 from earthlab/master

update

nkorinek committed:
Legend raster assertion fix!

nkorinek committed:
Changelog and black

nkorinek committed:
Additional vector tests (#212)

* Added a get_points() and assert_points() function to the vector tester.

* Added in get_points() and assert_points() functions with tests

* Added proper documentation

* Small fix to please codacy

* black

* Updated changelog

* First round of tests for vector

* black

* Rough draft for bug fix

* Fixed bug with multiple geometries plotted alongside bug with identical x values causing failure!

* typo

* Fixed small bug with markersize

* Added comments explaining code

* rough drafts of more tests

* Added more tests for legends

* Added more tests, and took out broken tests

* small codacy fix

* Fixed test!

* Taking out redundant tests

* Took out unneccesary tests

* Added plt.close

* Added more tests

* Typo

* more tests for uncovered parts of the vector file

* black

* Fixed issues with vector checking truth value of dataframe, and added tests

* Fixing how vector checks for truth value of a dataframe

* Added more coverage!

* Broke tests up into individual files

* black

* Added tests for outlier cases

* Update matplotcheck/tests/test_points.py

typo

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* took plt.gca() out of tests

* Added changes suggested on GitHub

* Update CHANGELOG.md

* fix import order

* import order fix

* import order

* import order

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

nkorinek committed:
Merge pull request #14 from earthlab/master

update

nkorinek committed:
Fixed merge conflicts

nkorinek committed:
Add tests to the Autograde module (#224)

* Added tests for the autograde module

* removed unneccesary import

* Fixing issues that were failing in base_data

* codacy

* flake8

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Updated changelog

* flake8 fix

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

nkorinek committed:
Merge pull request #15 from earthlab/master

Add tests to the Autograde module (#224)

nkorinek committed:
Hist bin midpoints (#204)

* Add a get and assert function for bin midpoints, as well as tests for those functions

* Updated docstrings

* Added changelog changes and fixed minor formatting issue

* small flake8 fix

* Added midpoints functionatlity to the vignette for histogram testing

* Adding in changes suggested on GitHub1

* Fixed function description

* Took out all instances of  in example

* update docs

* Fixed small bug with section titles

* Update matplotcheck/tests/test_base_data.py

* reworded changelog

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

Commit:
Merge branch 'master' into raster-legend-fix

pyup-bot committed:
Update setuptools from 46.1.1 to 46.1.3 (#231)

nkorinek committed:
Merge pull request #16 from earthlab/master

update

nkorinek committed:
Title assert accepts strings (#229)

* Fixed assert_title_contains to take strings as well as lists

* Better implementation

* Changelog update

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

Commit:
Merge branch 'master' into raster-legend-fix

lwasser commented:
@nkorinek i just slacked you - if you can please clean up the gist and clearly show me how this works / doesn't work. I think we need to capture some errors here and return more useful information to the user. so let's use that gist as a way to clean this up just a bit so i can see how it works and suggest how i think it should work in terms of user input.

lwasser reviewed (changes requested):
Need some clarification of the gist!

nkorinek committed:
Merge pull request #17 from earthlab/master

Title assert accepts strings (#229)

nkorinek committed:
merge

nkorinek committed:
Cleared up errors thrown

nkorinek committed:
Merge branch 'raster-legend-fix' of https://github.com/nkorinek/matplotcheck into raster-legend-fix

nkorinek commented:
@lwasser  updated gist here! https://gist.github.com/nkorinek/e8aa5ec22cdaa8cca4acd2789f6031ff if you run it locally the error thrown will be different since I updated them to make more since.

lwasser reviewed (changes requested):
ok @nkorinek i'm going to leave some more specific feedback here but i'd like to do a few things
1. the assert function name is super long. let's make is shorter - oi'll make some suggestions in the code
2. since we are doing the work to grab the legend labels, we should make a method that does it rather than having all of the code in the assert statement. this will be cleaner, easier to test and manage. i'll leave more specifics here in a minute..

lwasser commented on matplotcheck/raster.py:
```diff
@@ -140,25 +140,36 @@ def assert_legend_accuracy_classified_image(
         assert len([val for val in legend_dict.values() if val]) == len(
             all_label_options
-        ), "Incorrect legend labels"
+        ), (
```
let's rename this from
assert_legend_accuracy_classified_image
to 
assert_legend_labels 

i'm not sure why that other name was selected but it seems confusing!

lwasser commented on matplotcheck/raster.py:
```diff
@@ -140,25 +140,36 @@ def assert_legend_accuracy_classified_image(
             all_label_options
-        ), "Incorrect legend labels"
+        ), (
+            "Number of label options provided doesn't match the number of",
```
@nkorinek i can't comment inline here but i'd like to see this method broken up into pieces we can use and test independently.

1. get_plot_image or get_image. i suspect this method already exists and we can just reuse it here to grab the array rather than rewriting it which creates redundancy in the code
2. get_legend labels -- that could be a nice helper that we can look at and test.
3. finally the assert itself which compares an input list to the expected values.

lwasser commented on matplotcheck/raster.py:
```diff
@@ -140,25 +140,36 @@ def assert_legend_accuracy_classified_image(
         ]
         im_expected_labels = [
-            [all_label_options[val][0] for val in row] for row in im_expected
+            [all_label_options[val - correction][0] for val in row]
```
just a note - this method is super long and i think the parts of it could be modulearized and some reused elsewhere (get_image) for example. so let's clean this up.

I see that the notebook is working but i think we should clean up this method and then look at how is works more closely but looking at the various methods! 

note: CI is failing now but it's an known issue and a new one so we can keep working and assume it will fail. if i need to remove m2r we can consider that... i left a comment about m2r and the failure in my most recent pr to update docs!

lwasser commented:
@nkorinek i see a LOT of merge master commits in this pr. i suspect we need to find a way for you to update from master  at the CLI or rebase...  let's try a few things for your next few pr's

nkorinek committed:
Beginning to create more functions

nkorinek committed:
Moving parts of assert_legends into different helper functions to make the function shorter

nkorinek commented:
This should be ready for a review btw @lwasser . The docs are only failing b/c of that m2r bug.

nkorinek committed:
M2r remove (#247)

* Removed m2r from matplotcheck, and reformated code of conduct

* Seeing if rebase works

* removed an m2r import

* Fixing up the docs a bit

* changelog updates

nkorinek committed:
Merge pull request #18 from earthlab/master

M2r remove (#247)

nkorinek committed:
merge

nkorinek committed:
merge

nkorinek committed:
Updating in order to squash commits!

lwasser commented:
ok @nkorinek sorry to do this to you. this has a conflict and needs an update from master to fix m2r. i'll have a look again when ci is passing. you may want to check in on some of your other pr's as well as they likely all need updates.

lwasser commented on matplotcheck/raster.py:
```diff
@@ -114,19 +112,65 @@ def assert_legend_accuracy_classified_image(
+
+        # IMAGE TESTS/HELPER FUNCTIONS
+
+    def get_legend_labels(self, all_label_options):
```
@nkorinek this helper is confusing. i expected it to return the labels that it found in the existing legend. But it seems to check the labels? It seems to me having a helper that just returned the labels would be really helpful. then you can have a separate helper that checks that the correct words are in each label element. does that make sense? this function name suggests it will just grab the legend labels which is ideal. i may be confused by what is happening here... so we can chat i fyou want

lwasser replied:
<img width="1154" alt="Screen Shot 2020-04-20 at 5 48 51 PM" src="https://user-images.githubusercontent.com/7649194/79810020-3be84200-832f-11ea-85de-3b7206f058f4.png">
i'm getting some odd outputs from this method. does this make sense to you?

lwasser replied:
i am guessing that could be rgb values and then data value exept the 4th value is always 1.0.

lwasser replied:
```
def assert_legend_labels(expected_labels):
     self.get_legend_labels()
     # check stuff
     # Make sure that the expected labels are in the legend labels
```

lwasser commented on matplotcheck/raster.py:
```diff
@@ -140,27 +184,12 @@ def assert_legend_accuracy_classified_image(
         # Check that each legend entry label is in one of all_label_options
         assert len([val for val in legend_dict.values() if val]) == len(
```
Here it looks like you're testing that the labels occur in the legend? it seems like this helper should just return the labels in the legend for the plot.

lwasser commented on matplotcheck/raster.py:
```diff
@@ -140,27 +184,12 @@ def assert_legend_accuracy_classified_image(
         assert len([val for val in legend_dict.values() if val]) == len(
             all_label_options
-        ), "Incorrect legend labels"
+        ), (
```
this test could then be in the "assert correct labels" method rather than here.

lwasser commented on matplotcheck/raster.py:
```diff
@@ -140,27 +184,12 @@ def assert_legend_accuracy_classified_image(
-        ), "Incorrect legend to data relation"
-
-        # IMAGE TESTS/HELPER FUNCTIONS
+        return legend_dict
```
so i'm curious about this part. i vaguely remember checking that the colormap matched the labels and wonder if that is still happening? i remember working with kristen on this but it was a long time ago.

lwasser commented on matplotcheck/raster.py:
```diff
@@ -140,27 +184,12 @@ def assert_legend_accuracy_classified_image(
-        # labels (im_data_labels) and the other with the expected labels
-        # (im_expected_labels)
-        im_class_dict = {}
-        for val in np.unique(im_data):
```
this is the part i'm wondering about - does this map cmap values to colors and labels?

lwasser reviewed (changes requested):
we may need to chat a bit about this guy @nkorinek . great effort on it. i am just confused about a few things in how this is constructed and what the methods do.

i think we want a few methods here to help us along
1. get_cmap -- something that grabs the colormap used to associated colors with values in the image array
2. get_legend_labels - returns just the labels from the legend and potentially the color associated with each label or that could be a separate method.
3. assert legend labels - checks that the labels are correct (purple maps to gain, etc). 
right now it seems as if we may need to chat a bit more about this !

Commit:
fix copyright on docs (#244)

lwasser commented on matplotcheck/raster.py:
```diff
@@ -80,9 +80,7 @@ def _which_label(self, label, all_label_options):
-    def assert_legend_accuracy_classified_image(
-        self, im_expected, all_label_options
-    ):
+    def assert_legend_labels(self, im_expected, all_label_options):
```
let's rename `assert_raster_legend_labels`

lwasser commented on matplotcheck/raster.py:
```diff
@@ -114,19 +112,65 @@ def assert_legend_accuracy_classified_image(
-            im = self.ax.get_images()[0]
-            im_data, im_cmap = im.get_array(), im.get_cmap()
+        im_data = self.get_plot_image()
+        im = self.ax.get_images()[0]
```
let's make a method called `get_image_cmap()` for this. It can be short.

lwasser commented on matplotcheck/raster.py:
```diff
@@ -114,19 +112,65 @@ def assert_legend_accuracy_classified_image(
+        im = self.ax.get_images()[0]
+        im_cmap = im.get_cmap()
+
         assert list(im_data), "No Image Displayed"
```
```suggestion
        assert list(im_data), "There is no image in this plot"
```

lwasser replied:
Maybe add a comment above that says - `if the plot is empty, return error`

lwasser replied:
get_plot_image() actually tests that an image exists. so i'm not sure why we even need this test here again. i'm still doubtful that this is the best way to test things but i do think a try: get_plot_image() is enough to ensure that this works. `try:` can caption the assertion error and fail gracefully.

We definitely want to add a few tests for plots with no image in them.

lwasser commented on matplotcheck/raster.py:
```diff
@@ -114,19 +112,65 @@ def assert_legend_accuracy_classified_image(
 
+        # Retrieve legend entries and find which element of all_label_options
+        # matches that entry
+        legend_dict = self.get_legend_labels(all_label_options)
```
why does this need all_label_options (and then let's rename all_label_options to `expected_labels`

Commit:
Add pillow (#254)

* add pillow as a dev requirement

* update for pillow addition

Commit:
Merge branch 'master' into raster-legend-fix

lwasser commented:
/rebase

nkorinek committed:
Added get_cmap function

nkorinek committed:
Merge

nkorinek committed:
Merge branch 'earthlab-master'

nkorinek committed:
merge

nkorinek committed:
Complete overhaul of raster legend labels check and updates to the testing around it

nkorinek committed:
merge

nkorinek commented:
@lwasser This is (finally) updated! Took a bit of time to figure out how the old tests were working, and how I wanted to change them. This is basically a complete overhaul of this function. I'd be happy to talk about the changes I made to the functions, I think it'd be good to bounce this idea off of someone else before it gets merged, but I think it's a big improvement over how this used to run! Let me know if you have any immediate questions.

nkorinek committed:
Minor wording updates

nkorinek committed:
Changelog to rst (#267)

* Update setuptools from 42.0.2 to 45.2.0 (#190)

* raster inherits from vector rather than base (#76)

* Allow geodataframes in assert_polygons (#188)

* fixing a few syntax errors (#194)

* First draft of get_images function

* Added a test for get_image

* Changelog update

* minor formatting change.

* black changes

* fixing a few syntax errors

* Update matplotcheck/raster.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* remove tabls

Co-authored-by: Nathan Korinek <nako1890@colorado.edu>

* update changelog for release

* Bump version: 0.1.1 → 0.1.2

* Add flake8 fix [WIP] (#199)

* add flake 8

* yay cleanup

* update change log

* add flake 8 dep

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* more line length issues

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* fixed test_base_axis.py

* fixed text_base_data.py

* fixed test_base_legends.py

* fixed test_base_titles_captions.py

* Format files to be flake8 acceptable (#197)

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* Fixed test_raster.py and issue with legends tests

* black

* more black changes

* Fixing merge conflicts that were found

* Changing accepted changes

* minor changes

* Changing to make CI happy, modified to pass black and make -B docs

* Reformatting to make black and flake8 happy. Also updated crs assignment

* Fixed formatting issue with timeseries

* Better fix for timeseries formatting

* Black

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

* Assert string accept spaces (#205)

* Made it so that assert_string_contains accepts key words with spaces

* black

* renaming function

* Showed assert_title_contains working with a space in the keyword in vignette

* Update codecov to 2.0.16 (#202)

* Update codecov from 2.0.15 to 2.0.16

* update from master (#211)

* Add flake8 fix [WIP] (#199)

* add flake 8

* yay cleanup

* update change log

* add flake 8 dep

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* more line length issues

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* fixed test_base_axis.py

* fixed text_base_data.py

* fixed test_base_legends.py

* fixed test_base_titles_captions.py

* Format files to be flake8 acceptable (#197)

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* Fixed test_raster.py and issue with legends tests

* black

* more black changes

* Fixing merge conflicts that were found

* Changing accepted changes

* minor changes

* Changing to make CI happy, modified to pass black and make -B docs

* Reformatting to make black and flake8 happy. Also updated crs assignment

* Fixed formatting issue with timeseries

* Better fix for timeseries formatting

* Black

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

* Assert string accept spaces (#205)

* Made it so that assert_string_contains accepts key words with spaces

* black

* renaming function

* Showed assert_title_contains working with a space in the keyword in vignette

Co-authored-by: Nathan Korinek <nako1890@colorado.edu>

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>
Co-authored-by: Nathan Korinek <nako1890@colorado.edu>

* Contributors update (#213)

* Added contributor rst file and moved that information out of the README file

* Implemented requested changes to README

* Update setuptools from 45.2.0 to 46.0.0 (#215)

* Update pytest from 5.3.5 to 5.4.1

* Add Assert points function & cleanup duplicate methods (#203)

* Added a get_points() and assert_points() function to the vector tester.

* Added in get_points() and assert_points() functions with tests

* Added proper documentation

* Small fix to please codacy

* black

* Updated changelog

* Rough draft for bug fix

* Fixed bug with multiple geometries plotted alongside bug with identical x values causing failure!

* typo

* Fixed small bug with markersize

* Added comments explaining code

* Taking out redundant tests

* Typo

* Fixing how vector checks for truth value of a dataframe

* merge

* Additional vector tests (#212)

* Added a get_points() and assert_points() function to the vector tester.

* Added in get_points() and assert_points() functions with tests

* Added proper documentation

* Small fix to please codacy

* black

* Updated changelog

* First round of tests for vector

* black

* Rough draft for bug fix

* Fixed bug with multiple geometries plotted alongside bug with identical x values causing failure!

* typo

* Fixed small bug with markersize

* Added comments explaining code

* rough drafts of more tests

* Added more tests for legends

* Added more tests, and took out broken tests

* small codacy fix

* Fixed test!

* Taking out redundant tests

* Took out unneccesary tests

* Added plt.close

* Added more tests

* Typo

* more tests for uncovered parts of the vector file

* black

* Fixed issues with vector checking truth value of dataframe, and added tests

* Fixing how vector checks for truth value of a dataframe

* Added more coverage!

* Broke tests up into individual files

* black

* Added tests for outlier cases

* Update matplotcheck/tests/test_points.py

typo

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* took plt.gca() out of tests

* Added changes suggested on GitHub

* Update CHANGELOG.md

* fix import order

* import order fix

* import order

* import order

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

* Add tests to the Autograde module (#224)

* Added tests for the autograde module

* removed unneccesary import

* Fixing issues that were failing in base_data

* codacy

* flake8

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Updated changelog

* flake8 fix

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

* Hist bin midpoints (#204)

* Add a get and assert function for bin midpoints, as well as tests for those functions

* Updated docstrings

* Added changelog changes and fixed minor formatting issue

* small flake8 fix

* Added midpoints functionatlity to the vignette for histogram testing

* Adding in changes suggested on GitHub1

* Fixed function description

* Took out all instances of  in example

* update docs

* Fixed small bug with section titles

* Update matplotcheck/tests/test_base_data.py

* reworded changelog

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

* Update setuptools from 46.1.1 to 46.1.3 (#231)

* Title assert accepts strings (#229)

* Fixed assert_title_contains to take strings as well as lists

* Better implementation

* Changelog update

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

* M2r remove (#247)

* Removed m2r from matplotcheck, and reformated code of conduct

* Seeing if rebase works

* removed an m2r import

* Fixing up the docs a bit

* changelog updates

* # This is a combination of 2 commits.
# This is the 1st commit message:

Update setuptools from 42.0.2 to 45.2.0 (#190)

raster inherits from vector rather than base (#76)

Allow geodataframes in assert_polygons (#188)

fixing a few syntax errors (#194)

* First draft of get_images function

* Added a test for get_image

* Changelog update

* minor formatting change.

* black changes

* fixing a few syntax errors

* Update matplotcheck/raster.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* remove tabls

Co-authored-by: Nathan Korinek <nako1890@colorado.edu>

update changelog for release

Bump version: 0.1.1 → 0.1.2

Add flake8 fix [WIP] (#199)

* add flake 8

* yay cleanup

* update change log

* add flake 8 dep

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* more line length issues

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* fixed test_base_axis.py

* fixed text_base_data.py

* fixed test_base_legends.py

* fixed test_base_titles_captions.py

* Format files to be flake8 acceptable (#197)

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* Fixed test_raster.py and issue with legends tests

* black

* more black changes

* Fixing merge conflicts that were found

* Changing accepted changes

* minor changes

* Changing to make CI happy, modified to pass black and make -B docs

* Reformatting to make black and flake8 happy. Also updated crs assignment

* Fixed formatting issue with timeseries

* Better fix for timeseries formatting

* Black

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

Assert string accept spaces (#205)

* Made it so that assert_string_contains accepts key words with spaces

* black

* renaming function

* Showed assert_title_contains working with a space in the keyword in vignette

Update codecov to 2.0.16 (#202)

* Update codecov from 2.0.15 to 2.0.16

* update from master (#211)

* Add flake8 fix [WIP] (#199)

* add flake 8

* yay cleanup

* update change log

* add flake 8 dep

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* more line length issues

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* fixed test_base_axis.py

* fixed text_base_data.py

* fixed test_base_legends.py

* fixed test_base_titles_captions.py

* Format files to be flake8 acceptable (#197)

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* Fixed test_raster.py and issue with legends tests

* black

* more black changes

* Fixing merge conflicts that were found

* Changing accepted changes

* minor changes

* Changing to make CI happy, modified to pass black and make -B docs

* Reformatting to make black and flake8 happy. Also updated crs assignment

* Fixed formatting issue with timeseries

* Better fix for timeseries formatting

* Black

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

* Assert string accept spaces (#205)

* Made it so that assert_string_contains accepts key words with spaces

* black

* renaming function

* Showed assert_title_contains working with a space in the keyword in vignette

Co-authored-by: Nathan Korinek <nako1890@colorado.edu>

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>
Co-authored-by: Nathan Korinek <nako1890@colorado.edu>

Contributors update (#213)

* Added contributor rst file and moved that information out of the README file

* Implemented requested changes to README

Update setuptools from 45.2.0 to 46.0.0 (#215)

Add Assert points function & cleanup duplicate methods (#203)

* Added a get_points() and assert_points() function to the vector tester.

* Added in get_points() and assert_points() functions with tests

* Added proper documentation

* Small fix to please codacy

* black

* Updated changelog

* Rough draft for bug fix

* Fixed bug with multiple geometries plotted alongside bug with identical x values causing failure!

* typo

* Fixed small bug with markersize

* Added comments explaining code

* Taking out redundant tests

* Typo

* Fixing how vector checks for truth value of a dataframe

Update pytest from 5.3.5 to 5.4.1

merge

Additional vector tests (#212)

* Added a get_points() and assert_points() function to the vector tester.

* Added in get_points() and assert_points() functions with tests

* Added proper documentation

* Small fix to please codacy

* black

* Updated changelog

* First round of tests for vector

* black

* Rough draft for bug fix

* Fixed bug with multiple geometries plotted alongside bug with identical x values causing failure!

* typo

* Fixed small bug with markersize

* Added comments explaining code

* rough drafts of more tests

* Added more tests for legends

* Added more tests, and took out broken tests

* small codacy fix

* Fixed test!

* Taking out redundant tests

* Took out unneccesary tests

* Added plt.close

* Added more tests

* Typo

* more tests for uncovered parts of the vector file

* black

* Fixed issues with vector checking truth value of dataframe, and added tests

* Fixing how vector checks for truth value of a dataframe

* Added more coverage!

* Broke tests up into individual files

* black

* Added tests for outlier cases

* Update matplotcheck/tests/test_points.py

typo

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* took plt.gca() out of tests

* Added changes suggested on GitHub

* Update CHANGELOG.md

* fix import order

* import order fix

* import order

* import order

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

Add tests to the Autograde module (#224)

* Added tests for the autograde module

* removed unneccesary import

* Fixing issues that were failing in base_data

* codacy

* flake8

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Update matplotcheck/tests/test_autograde.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* Updated changelog

* flake8 fix

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

Hist bin midpoints (#204)

* Add a get and assert function for bin midpoints, as well as tests for those functions

* Updated docstrings

* Added changelog changes and fixed minor formatting issue

* small flake8 fix

* Added midpoints functionatlity to the vignette for histogram testing

* Adding in changes suggested on GitHub1

* Fixed function description

* Took out all instances of  in example

* update docs

* Fixed small bug with section titles

* Update matplotcheck/tests/test_base_data.py

* reworded changelog

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

Update setuptools from 46.1.1 to 46.1.3 (#231)

Title assert accepts strings (#229)

* Fixed assert_title_contains to take strings as well as lists

* Better implementation

* Changelog update

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

M2r remove (#247)

* Removed m2r from matplotcheck, and reformated code of conduct

* Seeing if rebase works

* removed an m2r import

* Fixing up the docs a bit

* changelog updates

Allow geodataframes in assert_polygons (#188)

fixing a few syntax errors (#194)

* First draft of get_images function

* Added a test for get_image

* Changelog update

* minor formatting change.

* black changes

* fixing a few syntax errors

* Update matplotcheck/raster.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* remove tabls

Co-authored-by: Nathan Korinek <nako1890@colorado.edu>

Get images function (#193)

* First draft of get_images function

* Added a test for get_image

* Changelog update

* minor formatting change.

* black changes

* Update matplotcheck/raster.py

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

Bump version: 0.1.1 → 0.1.2

Add flake8 fix [WIP] (#199)

* add flake 8

* yay cleanup

* update change log

* add flake 8 dep

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* more line length issues

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* fixed test_base_axis.py

* fixed text_base_data.py

* fixed test_base_legends.py

* fixed test_base_titles_captions.py

* Format files to be flake8 acceptable (#197)

* autograde flake8 fixes

* Fixed base to be flake8 compliant (phew)

* Made folium.py flake8 compliant

* made notebook.py flake8 compliant

* Fixed raster.py for flake8

* Fixed timeseries.py

* fixed conftest.py

* fixed test_base.py

* Fixed test_raster.py and issue with legends tests

* black

* more black changes

* Fixing merge conflicts that were found

* Changing accepted changes

* minor changes

* Changing to make CI happy, modified to pass black and make -B docs

* Reformatting to make black and flake8 happy. Also updated crs assignment

* Fixed formatting issue with timeseries

* Better fix for timeseries formatting

* Black

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

Contributors update (#213)

* Added contributor rst file and moved that information out of the README file

* Implemented requested changes to README

Add Assert points function & cleanup duplicate methods (#203)

* Added a get_points() and assert_points() function to the vector tester.

* Added in get_points() and assert_points() functions with tests

* Added proper documentation

* Small fix to please codacy

* black

* Updated changelog

* Rough draft for bug fix

* Fixed bug with multiple geometries plotted alongside bug with identical x values causing failure!

* typo

* Fixed small bug with markersize

* Added comments explaining code

* Taking out redundant tests

* Typo

* Fixing how vector checks for truth value of a dataframe

Additional vector tests (#212)

* Added a get_points() and assert_points() function to the vector tester.

* Added in get_points() and assert_points() functions with tests

* Added proper documentation

* Small fix to please codacy

* black

* Updated changelog

* First round of tests for vector

* black

* Rough draft for bug fix

* Fixed bug with multiple geometries plotted alongside bug with identical x values causing failure!

* typo

* Fixed small bug with markersize

* Added comments explaining code

* rough drafts of more tests

* Added more tests for legends

* Added more tests, and took out broken tests

* small codacy fix

* Fixed test!

* Taking out redundant tests

* Took out unneccesary tests

* Added plt.close

* Added more tests

* Typo

* more tests for uncovered parts of the vector file

* black

* Fixed issues with vector checking truth value of dataframe, and added tests

* Fixing how vector checks for truth value of a dataframe

* Added more coverage!

* Broke tests up into individual files

* black

* Added tests for outlier cases

* Update matplotcheck/tests/test_points.py

typo

Co-Authored-By: Leah Wasser <leah.wasser@colorado.edu>

* took plt.gca() out of tests

* Added changes suggested on GitHub

* Update CHANGELOG.md

* fix import order

* import order fix

* import order

* import order

Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

# This is the commit message #2:

Update setuptools from 46.1.1 to 46.1.3 (#231)

* fix copyright on docs (#244)

* Add pillow (#254)

* add pillow as a dev requirement

* update for pillow addition

* Changed the changelog to an rst file instead of a markdown file and added it to the docs build.

* Put changelog changes into the new changelog lol

Co-authored-by: pyup.io bot <github-bot@pyup.io>
Co-authored-by: Leah Wasser <leah.wasser@colorado.edu>

Commit:
setup greetings! (#257)

nkorinek committed:
Update from master

nkorinek committed:
Merge branch 'master' into raster-legend-fix

lwasser commented on matplotcheck/raster.py:
```diff
@@ -57,14 +57,15 @@ def assert_colorbar_range(self, crange):
             ), "Colorbar maximum is not expected value:{0}".format(crange[1])
 
     def _which_label(self, label, all_label_options):
```
let's rename `all_label_options` to `expected_labels`. And let's make that a cnvention in MPC where - whenever we are using parameters that ask for expected values we name them `expected_name_of_thing_here`. this will make the code generally a lot easier to follow!!

lwasser replied:
let's also rename this helper -- > `_check_label_matches_expected`
and have this return a boolean of True if it matches and false if it doesnt

lwasser commented on matplotcheck/raster.py:
```diff
@@ -57,14 +57,15 @@ def assert_colorbar_range(self, crange):
     def _which_label(self, label, all_label_options):
-        """Helper function for assert_legend_accuracy_classified_image
+        """Helper function for assert_legend_labels
         Returns string that represents a category label for label.
```
This is a very confusing docstring! it seems like this helper actually tests to see which input labels (from the student plot) match the expected labels (from the instructor test). so let's clarify both the helper function name and the docstring. 

_check_labels or something like that? 

also the output of this method is odd. it returns None if the labels DO MATCH (i think??) and the label that is incorrect if not. My suggestion is that it returns a boolean that can be used to then pull out which provided labels from the student plot are incorrect.

here are my notes -- see what you think about this approach.
```
    labels_check = [_which_label(label, all_label_options[i]) for i, label
        in enumerate(labels)]

    # Check that each legend entry label is in one of all_label_options
    # So it might be useful here to atleast print out the label found that
    # does NOT match the expected label. if which_labels returns a boolean i
    # think you could get those values by doing labels[boolean array]. and
    # then print this out.
```

lwasser commented on matplotcheck/raster.py:
```diff
@@ -57,14 +57,15 @@ def assert_colorbar_range(self, crange):
         label: string from legend to see if it contains an option in
             all_label_options
-        all_label_options: list of lists
+        all_label_options: list
```
expected_labels

lwasser commented on matplotcheck/raster.py:
```diff
@@ -57,14 +57,15 @@ def assert_colorbar_range(self, crange):
 
         Parameters
         ----------
         label: string from legend to see if it contains an option in
```
this is confusing. i think label is  a single label found in the input plot legend.

lwasser commented on matplotcheck/raster.py:
```diff
@@ -74,22 +75,18 @@ def _which_label(self, label, all_label_options):
         string that is the first entry in the internal list which label is
         matched with. If no match is found, return value is None
```
this just seems confusing. why not just return a boolean list and clarify what this helper does.

lwasser commented on matplotcheck/raster.py:
```diff
@@ -74,22 +75,18 @@ def _which_label(self, label, all_label_options):
-            for s in label_opts:
-                if s in label:
-                    return label_opts[0]
+        for label_option in all_label_options:
```
i definitely think we need more tests. 
let's have a fixture that as two legends for a plot and see how it works with the changes as is. i suspect that it could be problematic but let's test it out and then see what we need to do to accomodate legends that are more complex.  just create a plot with two legends somewhat like we do with geopandas objects and see how this works!!

lwasser commented on matplotcheck/raster.py:
```diff
@@ -74,22 +75,18 @@ def _which_label(self, label, all_label_options):
-        self, im_expected, all_label_options
-    ):
+    def assert_raster_legend_labels(self, im_expected, all_label_options):
         """Asserts legend correctly describes classified image on Axes ax,
```
Tests whether the legend labels on the input plot match labels that are expected.

lwasser commented on matplotcheck/raster.py:
```diff
@@ -100,68 +97,62 @@ def assert_legend_accuracy_classified_image(
-        all_label_options). The same is done for the expected image array.
-        Finally those two arrays of strings are compared. Passes if they match.
         """
         # Retrieve image array
```
looking at kristen's original description... i suspect that some functionality that she had included has been removed here... so i am going to go back to her original code. specifically we want to make sure that the patches and associated labels n the plot map to the expected array correctly. ... more in a bit on this.

lwasser commented on matplotcheck/raster.py:
```diff
@@ -100,68 +97,62 @@ def assert_legend_accuracy_classified_image(
+        ), "Provided legend labels don't match labels found."
+
+        # Check that expected and actual arrays data match up
         assert np.array_equal(
```
this test really doesn't get at what we need to do. it just tests that the image is correct which we do elsewhere. so it probably can be removed.

lwasser commented on matplotcheck/raster.py:
```diff
@@ -100,68 +97,62 @@ def assert_legend_accuracy_classified_image(
 
-        # Check that expected and actual labels match up
+        # Check that each legend entry label is in one of all_label_options
+        assert all(
```
here it would be more useful to say xxx label(s) were incorrect -- print out which labels were found to be incorrect.

lwasser commented on matplotcheck/raster.py:
```diff
@@ -100,68 +97,62 @@ def assert_legend_accuracy_classified_image(
 
         # IMAGE TESTS/HELPER FUNCTIONS
 
+    def get_legend_labels(self):
```
i think we need to reorder methods - it doesn't make a lot of sense to call methods before they are defined even tho it works with classes. it would be better to organize things so the helpers come BEFORE the main method that does all of the checks.

lwasser commented on matplotcheck/raster.py:
```diff
@@ -100,68 +97,62 @@ def assert_legend_accuracy_classified_image(
+            for label in sublist
+        ]
+
     def get_plot_image(self):
```
move this up - and then make sure all methods are defined before they are called in this module.

lwasser commented:
so this code
```
        # matches that entry
        legend_dict = {}
        for p in [
            p
            for sublist in [leg.get_patches() for leg in legends]
            for p in sublist
        ]:
            label = p.get_label().lower()
            legend_dict[p.get_facecolor()] = _which_label(
                label, all_label_options
            )
```
creates a dictionary with 
color: mapped to the output of which label. 
i thijnk we definitely want an output dictionary that pulls the students plot legend 
legend_label: "color found". And what we could do is then have another item in the dictionary representing whether the label is True (matches an expected label) or False (does not match an expected label. 

If which_label is renamed to be `_check_label` and then the docstring explains that this helper does a check to ensure the label in the student plot is found in the expected label values, that would be valuable. Once we have that we can then test that the color of the legend - label matches the value we expect in the image array. i think this will work well.

lwasser commented on matplotcheck/raster.py:
```diff
@@ -100,68 +97,62 @@ def assert_legend_accuracy_classified_image(
 
         # Retrieve legend entries and find which element of all_label_options
         # matches that entry
-        legend_dict = {}
```
ok we need this functionality but it could be a little helper method which would be nice that returns a dictionary. i'd rather the dictionary be setup with the plot labels as the KEY, then the values would be 1. color and 2. true  or false so false would be if it doesn't match an expected label value and true means it matches an expected label value.
